bazel_codegen/lib/src/run_phases.dart

```diff
@@ -0,0 +1,249 @@
+    if (inputPath.startsWith(outputDir)) {
+      return p.relative(inputPath, from: outputDir);
+    }
+  }
+  return inputPath;
+}
+
+/// Persistent worker loop implementation.
+class _CodegenWorker extends AsyncWorkerLoop {
+  final Map<String, String> defaultContent;
+  final PhaseFactory phaseFactory;
+  int numRuns = 0;
+
+  _CodegenWorker(this.phaseFactory, this.defaultContent);
+
+  @override
+  Future<WorkResponse> performRequest(WorkRequest request) async {
+    IOSinkLogHandle logHandle;
+    var options = _parseOptions(request.arguments);
+    try {
+      numRuns++;
+      var timings = new CodegenTiming()..start();
+
+      var bazelRelativeInputs = request.inputs
+          .map((input) => _bazelRelativePath(input.path, options.rootDirs));
+
+      // Actually run the phases
+      var phases = phaseFactory(options);
+      logHandle = await _runPhases(phases, options, defaultContent, timings,
+          isWorker: true, validInputs: new Set()..addAll(bazelRelativeInputs));
+      var logger = logHandle.logger;
+      logger.info(
+          'Completed in worker mode, this worker has ran $numRuns builds');
+      await logHandle.close();
+      var message = _loggerMessage(logHandle, options.logPath);
+
+      var response = new WorkResponse()
+        ..exitCode = logHandle.errorCount == 0 ? EXIT_CODE_OK : EXIT_CODE_ERROR;
+      if (message.isNotEmpty) response.output = message;
+      return response;
+    } catch (e, s) {
+      await logHandle?.close();
+      return new WorkResponse()
+        ..exitCode = EXIT_CODE_ERROR
+        ..output = "Dart Codegen worker failed with:\n$e\n$s";
+    }
+  }
+}
+
+/// Runs [phases] to generate files using [options].
+///
+/// The [timings] instance must already be started.
+///
+/// See one of the functions above for a description of
+/// [defaultContent].
+Future<IOSinkLogHandle> _runPhases(List<Phase> phases, Options options,
+    Map<String, String> defaultContent, CodegenTiming timings,
+    {bool isWorker: false, Set<String> validInputs}) async {
+  assert(timings.isRunning);
+
+  final srcPaths = await timings.trackOperation('Collecting input srcs', () {
+    return new File(options.srcsPath).readAsLines();
+  });
+  if (srcPaths.isEmpty) {
+    throw new CodegenError('No input files to process.');
+  }
+
+  final packageMap =
+      await timings.trackOperation('Reading package map', () async {
+    var lines = await new File(options.packageMapPath).readAsLines();
+    return new Map<String, String>.fromIterable(
+        lines.map((line) => line.split(':')),
+        key: (l) => l[0],
+        value: (l) => l[1]);
+  });
+
+  final writer = new BazelAssetWriter(options.outDir, packageMap,
+      validInputs: validInputs);
+  final reader = new BazelAssetReader(
+      options.packagePath, options.rootDirs, packageMap,
+      assetFilter: new AssetFilter(validInputs, packageMap, writer));
+  final srcAssets = timings.trackOperation('Reading input srcs', () {
```

Comment:
This probably doesn't need to be timed, or the description should be changed at least. It isn't actually reading anything.